server/manager/src/main/java/org/apache/accumulo/manager/tableOps/goal/SetHostingGoal.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.manager.tableOps.goal;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.accumulo.core.client.AccumuloException;
+import org.apache.accumulo.core.client.admin.TabletHostingGoal;
+import org.apache.accumulo.core.clientImpl.ClientContext;
+import org.apache.accumulo.core.clientImpl.thrift.TableOperation;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.NamespaceId;
+import org.apache.accumulo.core.data.PartialKey;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.accumulo.core.dataImpl.thrift.TRange;
+import org.apache.accumulo.core.fate.Repo;
+import org.apache.accumulo.core.metadata.schema.Ample.TabletsMutator;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata.Location;
+import org.apache.accumulo.core.metadata.schema.TabletsMetadata;
+import org.apache.accumulo.manager.Manager;
+import org.apache.accumulo.manager.tableOps.ManagerRepo;
+import org.apache.accumulo.manager.tableOps.Utils;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SetHostingGoal extends ManagerRepo {
+
+  private static final long serialVersionUID = 1L;
+  private static final Logger LOG = LoggerFactory.getLogger(SetHostingGoal.class);
+
+  private final TableId tableId;
+  private final NamespaceId namespaceId;
+  private final TRange range;
+  private final TabletHostingGoal goal;
+
+  public SetHostingGoal(TableId tableId, NamespaceId namespaceId, TRange range,
+      TabletHostingGoal goal) {
+    this.tableId = tableId;
+    this.namespaceId = namespaceId;
+    this.range = range;
+    this.goal = goal;
+  }
+
+  @Override
+  public long isReady(long tid, Manager manager) throws Exception {
+    return Utils.reserveNamespace(manager, namespaceId, tid, false, true,
+        TableOperation.SET_HOSTING_GOAL)
+        + Utils.reserveTable(manager, tableId, tid, true, true, TableOperation.SET_HOSTING_GOAL);
+  }
+
+  @Override
+  public Repo<Manager> call(long tid, Manager manager) throws Exception {
+    final Range r = new Range(range);
+    final List<KeyExtent> extents = findExtentsForRange(manager.getContext(), tableId, r);
+    LOG.debug("Setting tablet hosting goal to {} requested for: {} ", goal, extents);
+    try (TabletsMutator mutator = manager.getContext().getAmple().mutateTablets()) {
+      extents.forEach(e -> mutator.mutateTablet(e).setHostingGoal(goal).mutate());
```

Comment:
We need a follow on issue to address concurrency issues.  Current fate ops handle that w/ table lock and detection of concurrent split.  However I am not sure if its worthwhile to write code for this new fate op to do all of that as we may replace that w/ conditional mutations in the future.

Reply:
Do splits happen without a FATE table lock?

Reply:
yeah splits can happen w/o a table lock

Reply:
I think the closest example is bulk import.  It makes rpcs to the tablet and this helps avoid concurrent problems with splits.  The tablet on the tablet server writes the data to the metadata table.  The fate op keeps scanning until it sees the data in all tablets in the range.  The table lock helps avoid concurrent problems with destructive operations like delete table. 
So it avoids problems with concurrent operations with a combo of repeated metadata scans, tablet RPCs, and a table lock.

Reply:
Right, so it likely makes sense to make this a follow-on task and *not* add more RPCs to the Tablet/TabletServer now.

Reply:
In the new model of how Accumulo might work it would write a conditional mutation that requires an absent operation.  Like 

https://github.com/keith-turner/accumulo/blob/f06a87be124ed51eb31a06ca15135ae07ae8c2f8/server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java#L86

This would ensure there is no concurrent merge, delete, or split running.   It would have keep scanning until all conditional mutations are successful.

Reply:
For this particular operation we also want to consider the case of concurrency with itself.  Like  two concurrent setTabletHostingGoal API calls that overlap the same range but set different goal states.

Reply:
> For this particular operation we also want to consider the case of concurrency with itself. Like two concurrent setTabletHostingGoal API calls that overlap the same range but set different goal states.

Doesn't the table lock prevent that?

Reply:
> Doesn't the table lock prevent that?

Yeah, a table write lock would.  It also prevents all other operations from running (except for splits).

Reply:
Thinking through how this might be handled using only conditional mutations (and no table locks), something like the following may work using multiple fate steps.  The algorithm tries to reserve tablets before making the changes inorder to detect concurrent collisions with other FATE transactions trying to set the goal on the same range.

The following is the first fate step.

```
while(true) {
   foreach tablet in range {
        Do a conditional mutation settting coulumn changingGoal=<Fate tx id> IF column changingGoal is absent
   }  

   if conditional mutations failed {
      if there are other fate tx also changingGoal
         if the other fate tx id is lower then ours then delete our changingGoal columns and wait for the other fate tx to go away
         if the other fate tx id is higher than our then wait for them to go away
   } else {
     break;
   }
}
```

The second fate step would set the goals.  When we get to this step, all of the tablets are reserved for changing the goal so can go ahead and do it.

```
while(true) {
   
   foreach tablet {
        Do a conditional mutation settting goal <Fate tx id> IF changing goal is set to my transaction id.
   }
  
   if conditional mutations failed {
     if changingGoal column is not set, then throw an error this is unexpected.
   } else {
      break;
   } 
}
```

The third fate step would delete the changingGoal columns in the range.